### ISSUE TITLE
Support `/` for dividing `fmpq` and/or rationals by other fmpz/rationals/integers

### ISSUE THREAD

**Issue by fingolfin** [oscar-system/Oscar.jl#1618]
This issue really spans multiple of our packages (at least AA, Nemo, Oscar; perhaps Hecke).

Compare this:
```
julia> 5//2 / 2
5//4

julia> 2 / (5//2)
```
with this:
```
julia> fmpq(5,2) / 5
ERROR: MethodError: no method matching /(::fmpq, ::Int64)
```

I see no reason why we don't match the behavior of Julia's `Rational` type for `/`. 
Basically I am asking for `/` methods for the following (and possibly more):
- `fmpq` by `fmpq`, `fmpz`, `<:Integer`
- `fmpq_poly` by `fmpq`, `fmpz`, `<:Integer`
- actually *any* polynomial (multi or univariate, any coefficient ring) by `fmpq`, `fmpz`, `<:Integer` would be nice
  - (If the input is e.g. a normed `fmpz_poly` and we divide by `2` then it'd be fine to return an error because the result is not an `fmpz_poly`; but I'd also be OK if this just was not supported for now, and only worked for `fmpq_poly` or polys with coefficients from a field

- `fmpq_mat` (or really any of our matrix types?) by `fmpq`, `fmpz`, `<:Integer`
- ... possibly more

Of course one can go beyond this and say it should be possible to divide a polynomial over a number field by elements of that number field

In the past I was told this did conflict with Julia's definition / documentation for `/`, but I don't see how? I also can't think of any inconsistency or confusion this might cause, at least in the case of `QQ`?

**Comment by thofma** [comment]
Since special cases for polynomials might make the system less consistent, I guess the question would be: What is the mathematical definition of `/` that we want to consistently use? I am against introducing `/` just rationals or the other objects for which this is proposed. My hope would be that if this is introduced, it should be introduced for all ring elements. I believe that we don't want to introduce nice hacks to make certain applications work or look nicer at the cost of consistency.

Secondly, I really like that for a binary operator `op`, elements `x, y` and some parent object `S`, we always have `S(op(x, y)) == op(S(x), S(y))` (equality also means that one side could error and the other not, e.g., things could become divisible etc.). So there is no weird/unexpected change of behaviour when coercing into an overstructure. The proposed mismatch between `2/3`, `ZZ(2)/ZZ(3)` and `Zx(2)/ZZ(2)` is not desirable in my opinion.

**Comment by fingolfin** [comment]
I have not proposed anything for `ZZ(2)/ZZ(3)` or `Zx(2)/ZZ(2)`. And I think there is a fundamental difference between defining `/` for arguments where the result "always" makes sense (except for division by zero, of course), i.e. if the left side is a field element or a polynomial or matrix with coefficient from a field.

Rules like `S(op(x, y)) == op(S(x), S(y))` are nice and I'd in fact be glad to enforce it -- IMHO dividing an `fmpq_poly` by a constant `fmpq_poly` should also be allowed, no problem there. But I think there is a careful balance between "nice rules" and being pragmatic. Julia makes a lot of choices for us that are highly problematic for us, such as how `1/2` and `2^100` are implemented. Indeed, somehow nobody seems to object to your rule being violated by `fmpz(2^100)` vs `fmpz(2)^fmpz(100)` ?

**Comment by fieker** [comment]
On Wed, Oct 12, 2022 at 06:28:16AM -0700, Max Horn wrote:
> I have not proposed anything for `ZZ(2)/ZZ(3)` or `Zx(2)/ZZ(2)`. And I think there is a fundamental difference between defining `/` for arguments where the result "always" makes sense (except for division by zero, of course), i.e. if the left side is a field element or a polynomial or matrix with coefficient from a field.
> 
> Rules like `S(op(x, y)) == op(S(x), S(y))` are nice and I'd in fact be
> glad to enforce it -- IMHO dividing an `fmpq_poly` by a constant
> `fmpq_poly` should also be allowed, no problem there. But I think
I would disagree here...

**Comment by thofma** [comment]
> I have not proposed anything for `ZZ(2)/ZZ(3)` or `Zx(2)/ZZ(2)`. And I think there is a fundamental difference between defining `/` for arguments where the result "always" makes sense (except for division by zero, of course), i.e. if the left side is a field element or a polynomial or matrix with coefficient from a field.

Maybe I misunderstand, but I thought you propose to define `Zx(2)/ZZ(2)` in your third bullet point.

> Rules like `S(op(x, y)) == op(S(x), S(y))` are nice and I'd in fact be glad to enforce it -- IMHO dividing an `fmpq_poly` by a constant `fmpq_poly` should also be allowed, no problem there. But I think there is a careful balance between "nice rules" and being pragmatic. Julia makes a lot of choices for us that are highly problematic for us, such as how `1/2` and `2^100` are implemented. Indeed, somehow nobody seems to object to your rule being violated by `fmpz(2^100)` vs `fmpz(2)^fmpz(100)` ?

Yes, this is bad, but why use this as a reason to add more inconsistencies?

If we introduce `/`, I think the best course of action would be a definition of what `x/y` is for say field (ring) elements with the same parent and then everything should derive from coercions (like we do now for `+`, `-` and `divexact`). Just adding methods for `/` where the right side is `fmpz` feels adhoc.

Can you confirm that all the examples in your original post would be solved by defining
```
/(x::T, y::T) where {T <: RingElem} = divexact(x, y)
```
with the additional methods derive from coercion, e.g., `/(x::fmpz_poly, y::fmpz) = /(x, parent(x)(y))` (let's say behaviour wise, as the implementation might be different)?

If this is the proposal, then I think this might be a good idea. The only objection one could raise for this would be that there is a subtle difference between `2/1 == 1.0` and `ZZ(2)/ZZ(1) == ZZ(1)`, which might trip people up.

**Comment by fingolfin** [comment]
> Can you confirm that all the examples in your original post would be solved by defining
>
>      /(x::T, y::T) where {T <: RingElem} = divexact(x, y)

In all examples I tried, `divexact` did exactly what I wanted! So I'd be happy to delegate `/` to `divexact` and let it behave more or less exactly like that would be good; so all the promotion and delegation and what not.

The above is not quite enough because I'd also want do be able to write `fmpq(5,2) / 3`. You mentioned that additional code may be needed. Not sure what you have in mind, but the following seems to work for me; the idea here being that we just delegate to `divexact`, and let promotion etc. there take care (so basically `/` works iff `divexact` works):
```julia
Base.:/(x::ModuleElem, y::RingElement) = divexact(x, y)
Base.:/(x::RingElem, y::RingElement) = divexact(x, y)
# for symmetry perhaps also the following (and then one more to
# resolve the ambiguity?
#Base.:/(x::RingElement, y::RingElem) = divexact(x, y)
#Base.:/(x::RingElem, y::RingElem) = divexact(x, y)
```
(Of course one could go beyond and even replace `RingElement` by `Any`; not sure if that's a good idea, though...)

Anyway, with those all examples I care about work:
```
julia> fmpq(5,3) / 2
5//6

julia> fmpq(5,3) / fmpz(2)
5//6

julia> R, x = QQ[:x]
(Univariate Polynomial Ring in x over Rational Field, x)

julia> (5*x) / 2
5//2*x

julia> (5*x) / fmpz(2)
5//2*x

julia> (5*x) / fmpq(10,3)
3//2*x

julia> M = matrix(QQ, [1 2 ; 3 4])
[1   2]
[3   4]

julia> M / 2
[1//2   1]
[3//2   2]

julia> M / fmpz(2)
[1//2   1]
[3//2   2]

julia> M / fmpq(10,3)
[3//10   3//5]
[9//10   6//5]
```

I am also OK with this raising an error:
```
julia> ZZ(2) / ZZ(3)
ERROR: ArgumentError: Not an exact division
```

But perhaps there are issues with those that I am missing?

**Comment by thofma** [comment]
No, this is all fine and expected. This is exactly the proposal that comes up from time to time.

The only question is whether we want the mismatch between `ZZ(1)/ZZ(1)` and `1/1`/`big"1"/big"1"`. Up to now we had declined and postponed the discussion for the future. But maybe the future is now? If people are in favour of this change, I am happy to go with it.

I remark here another thing that will come up. At first I thought it would allow us to change the `//` to `/` when printing, but our rational elements print as plain `x//y`. Changing the printing to `x/y` would make it not parse anymore (which would be a shame).

**Comment by fingolfin** [comment]
For me personally, the mismatch between `ZZ(1)/ZZ(1)` would be a feature, not a bug ;-). And perhaps the best argument I know why `fmpz` is not an `Integer`

Regarding the printing of rationals: well `ZZ(3)/ZZ(2)` would produce an error anyway, so that's another reason why it should keep using `//`, isn't it?

**Comment by thofma** [comment]
Ah oops, you are right about the printing of rationals. Not sure what I was thinking.

So yeah, if all are happy with this, why not.

**Comment by fingolfin** [comment]
So, where should this be added? In AbstractAlgebra.jl?

And what should be added? This:
```julia
Base.:/(x::ModuleElem, y::RingElement) = divexact(x, y)
Base.:/(x::RingElem, y::RingElement) = divexact(x, y)
```

But for symmetry perhaps also the following? Though I think this less often useful if at all? (the second method is to resolve the ambiguity)
```julia
Base.:/(x::RingElement, y::RingElem) = divexact(x, y)
Base.:/(x::RingElem, y::RingElem) = divexact(x, y)
```

**Comment by fingolfin** [comment]
Thinking about symmetry made me wonder if we shouldn't also do `Base.:\` -- and map that to `divexact_left`, perhaps? And then `Base.:/` should perhaps call `divexact_right` and not just `divexact`?

(That then lead to me reading the definitions to `divexact_left` and `divexact_right` and for `divexact` and then what these are defined as, and I am not quite happy about what I found, but that's going too far afield now... *sigh*)

**Comment by fieker** [comment]
On Fri, Oct 14, 2022 at 07:05:24AM -0700, Max Horn wrote:
> Thinking about symmetry made me wonder if we shouldn't also do `Base.:\` -- and map that to `divexact_left`, perhaps? And then `Base.:/` should perhaps call `divexact_right` and not just `divexact`?
> 
> (That then lead to me reading the definitions to `divexact_left` and `divexact_right` and for `divexact` and then what these are defined as, and I am not quite happy about what I found, but that's going too far afield now... *sigh*)
> 

I would rather be in favour of restricing / to the commutative world and
divexaxt. For the un-comm world people should be explicit....

**Comment by fieker** [comment]
On Fri, Oct 14, 2022 at 07:05:24AM -0700, Max Horn wrote:
> Thinking about symmetry made me wonder if we shouldn't also do `Base.:\` -- and map that to `divexact_left`, perhaps? And then `Base.:/` should perhaps call `divexact_right` and not just `divexact`?
> 
> (That then lead to me reading the definitions to `divexact_left` and `divexact_right` and for `divexact` and then what these are defined as, and I am not quite happy about what I found, but that's going too far afield now... *sigh*)
> 
Also, what about non-associativity?

**Comment by fingolfin** [comment]
I don't see any problems relating to non-associativity?

**Comment by thofma** [comment]
I have no strong opinion about the noncommutative setting. Just from looking at the name I cannot guess what `divexact_left`, `divexact_right`, `/` or `\` would do. The same confusion arises when guessing whether `divexact_left == \` or `divexact_left == /`.

I would lean to not implement `/` or `\` any of it unless someone asks for it or we notice that some of the generic code could be simpified greatly using `/`/`\`.

**Comment by thofma** [comment]
I had a quick chat with @fieker and the `/(x, y) = divexact(x, y)` seems like a good idea. To help the user, the error message for `/` (that is, `divexact`) should mention `//` for creating elements in the fraction field.

**Comment by fingolfin** [comment]
In light of recent discussions, maybe this is also relevant for @wdecker and others?

**Comment by wdecker** [comment]
The problem in our discussion is as follows: Given a ring R and a multiplicatively closed subset U of R, how do we create and/or print elements of the localization of R at U starting from pairs (p, q) in R x U. For example, can we enter p/q in Julia (printing elements as p/q as well), or use p//q (which seems not appropriate as this notation is also used for elements of quotient fields or ...

**Comment by lkastner** [comment]
Just to recap: I think we are all in favor of intuitive things like `fmpq(5,2) / 5` working, right? If not, I'd suggest having a poll on this. My take on this is that the learning curve for Oscar is already pretty steep, there are many places I regularly get confused (probably due to my polymake background...). I'd hope that we can just override the general definition for cases requiring a more sophisticated approach -- in the "worst" case we could just error and hint at what the user might mean?

**Comment by fingolfin** [comment]
Trying to make forward progress on this via https://github.com/Nemocas/AbstractAlgebra.jl/pull/1285. Feedback or help welcome

**Comment by fingolfin** [comment]
Resolved by AA 0.28.4
